In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# import the pakage
# read the datqa
# divide the numerical and charector

In [3]:
file_path=r"C:\Users\amol\Hydrabad\Naresh_IT\Data_Science&Ai_by_OmkarSir\Classes\EDA\DataFrame\Visadataset.csv"
df_visa=pd.read_csv(file_path)
df_visa

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [4]:
num=df_visa.select_dtypes(exclude='object').columns
num


Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object')

In [5]:
df_visa['prevailing_wage']

0           592.2029
1         83425.6500
2        122996.8600
3         83434.0300
4        149907.3900
            ...     
25475     77092.5700
25476    279174.7900
25477    146298.8500
25478     86154.7700
25479     70876.9100
Name: prevailing_wage, Length: 25480, dtype: float64

In [6]:
wage_data=df_visa['prevailing_wage']
Count=len(wage_data)
Min=np.min(wage_data)
Max=np.max(wage_data)
Mean=np.mean(wage_data)
Median=np.median(wage_data)

In [7]:
wage_data=df_visa['prevailing_wage']
Count=len(wage_data)
Min=round(np.min(wage_data),2)
Max= round(np.max(wage_data),2)
Mean= round(np.mean(wage_data),2)
Median= round(np.median(wage_data),2)
index=['Count','Min','Mean','Median','Max']
Values=[Count,Min,Mean,Median,Max]
pd.DataFrame(Values,index=index,columns=['prevailing_wage'])

prevailing_wage
Count          25480.00
Min                2.14
Mean           74455.81
Median         70308.21
Max           319210.27

In [8]:
wage_data=df_visa['prevailing_wage']
Count=len(wage_data)
Min=round(np.min(wage_data),2)
Max= round(np.max(wage_data),2)
Mean= round(np.mean(wage_data),2)
Median= round(np.median(wage_data),2)
index=['Count','Min','Mean','Median','Max']
Values=[Count,Min,Mean,Median,Max]
pd.DataFrame(Values,index=index,columns=['prevailing_wage'])

prevailing_wage
Count          25480.00
Min                2.14
Mean           74455.81
Median         70308.21
Max           319210.27

In [9]:
wage_data=df_visa['prevailing_wage']
Count=len(wage_data)
Min=round(np.min(wage_data),2)
Max= round(np.max(wage_data),2)
Mean= round(np.mean(wage_data),2)
Median= round(np.median(wage_data),2)
index=['Count','Min','Mean','Median','Max']
Values=[Count,Min,Mean,Median,Max]
pd.DataFrame(Values,index=index,columns=['prevailing_wage'])

prevailing_wage
Count          25480.00
Min                2.14
Mean           74455.81
Median         70308.21
Max           319210.27

In [10]:
l=[]
for i in num:
    data=df_visa['prevailing_wage']
    Count=len(data)
    Min=round(np.min(data),2)
    Max= round(np.max(data),2)
    Mean= round(np.mean(data),2)
    Median= round(np.median(data),2)
    p_25=round(np.percentile(data,25),)
    p_50=round(np.percentile(data,50),)
    p_75=round(np.percentile(data,75),)
    index=['Count','Min','Mean','Median',
           '25p','50p','75p','Max']
    Values=[Count,Min,Mean,Median,p_25,p_50,p_75,Max]
    pd.DataFrame(Values,index=index,columns=['prevailing_wage'])
    l.append(Values)

In [11]:
pd.DataFrame(l,index=num,columns=index).T

no_of_employees  yr_of_estab  prevailing_wage
Count          25480.00     25480.00         25480.00
Min                2.14         2.14             2.14
Mean           74455.81     74455.81         74455.81
Median         70308.21     70308.21         70308.21
25p            34015.00     34015.00         34015.00
50p            70308.00     70308.00         70308.00
75p           107736.00    107736.00        107736.00
Max           319210.27    319210.27        319210.27

In [12]:
df_visa.describe()

no_of_employees   yr_of_estab  prevailing_wage
count     25480.000000  25480.000000     25480.000000
mean       5667.043210   1979.409929     74455.814592
std       22877.928848     42.366929     52815.942327
min         -26.000000   1800.000000         2.136700
25%        1022.000000   1976.000000     34015.480000
50%        2109.000000   1997.000000     70308.210000
75%        3504.000000   2005.000000    107735.512500
max      602069.000000   2016.000000    319210.270000

**percentile_quantile**

In [14]:
40 p=200
40 perecentage of total data

SyntaxError: invalid syntax (607116054.py, line 1)

In [ ]:
np.percentile(wage_data,50)


In [ ]:
np.quantile(wage_data,0.5)

In [ ]:
np.median(wage_data)

- 50p=70308

- 50 perecentage of total data has less than 70308.20

- 50 perecentage of 25480 has less than 70309.20

- 12740 applicants ha wages less than 70309.20

In [ ]:
wage_data=df_visa['prevailing_wage']
p_50=np.percentile(wage_data,50)
wage_data<p_50

In [ ]:
wage_data=df_visa['prevailing_wage']
p_50=np.percentile(wage_data,50)
con=wage_data<p_50
len(df_visa[con])
    

In [ ]:
wage_data=df_visa['prevailing_wage']
p_25=np.percentile(wage_data,25)
con=wage_data<p_50
len(df_visa[con])==25*25480/100

In [ ]:
wage_data=df_visa['prevailing_wage']
p_75=np.percentile(wage_data,75)
con=wage_data<p_50
len(df_visa[con])==25*25480/100

In [ ]:
print('Count is ',Count)

In [ ]:
print('Minimum is',Min)

In [ ]:
print('Maximasi is',)

**histogram**

In [ ]:
wage_data=df_visa['prevailing_wage']

In [ ]:
wage_data.mean()==wage_data.median()

In [ ]:
step-1: calculate wagemean=u
step-2: calculate std of wage sigma
step-3: lb=u-1*sigma
step-3: ub=u-1*sigma
step-5: con1=wage_data>lb
step-5: con2=wage_data<ub
step-7:con=con1&con2
step-8: len(visa_df[con])



In [ ]:
wage_mean=wage_data.mean()
wag_std=wage_data.std()
lb=wage_mean-1*wag_std
ub=wage_mean+1*wag_std
con1=wage_data>lb
con2=wage_data<ub
con=con1&con2
len(df_visa[con])==68*len(df_visa)/100


In [ ]:
def eperical(k,n):
    wage_mean=wage_data.mean()
    wag_std=wage_data.std()
    lb=wage_mean-k*wag_std
    ub=wage_mean+n*wag_std
    con1=wage_data>lb
    con2=wage_data<ub
    con=con1&con2
    print(len(df_visa[con])==n*len(df_visa)/100)
eperical(2,95)


In [ ]:
plt.hist(wage_data)

In [ ]:
plt.hist(wage_data,bins=40)

In [ ]:
plt.hist(wage_data,bins=40)

- Data skewed

- Not follwing Normal distribution

- Means outliers are present

**Outliers analysis**  

box-plot

In [ ]:
plt.boxplot(wage_data)

In [ ]:
plt.boxplot(wage_data)
plt.show()

In [ ]:
plt.subplot(2,1,1).hist(wage_data)
plt.subplot(2,1,2).boxplot(wage_data,vert=False)
plt.show()

**Outliears data**

In [ ]:
 code-block:: none

          Q1-1.5IQR   Q1   median  Q3   Q3+1.5IQR
                       |-----:-----|
       o      |--------|     :     |--------|    o  o
                       |-----:-----|
     flier             <----------->            fliers
                            IQR

In [ ]:
q1=np.percentile(wage_data,25)
q3=np.percentile(wage_data,75)
IQR =q3-q1
lb=q1-1.5*IQR
ub=q3+1.5*IQR
con1=wage_data<lb
con2=wage_data>ub
con= con1 |con2
outliersdf=df_visa[con]
outliersdf




In [ ]:
q1=np.percentile(wage_data,25)
q3=np.percentile(wage_data,75)
IQR =q3-q1
lb=q1-1.5*IQR
ub=q3+1.5*IQR
con1=wage_data>lb
con2=wage_data<ub
con= con1 & con2
nontliersdf=df_visa[con]
nontliersdf

In [ ]:
d=outliersdf['prevailing_wage']
plt.subplot(2,1,1).hist(d)
plt.subplot(2,1,2).boxplot(d,vert=False)
plt.show()

In [ ]:
aa=plt.boxplot(wage_data)
aa.q1

In [ ]:
lb=2.13670000e+00
ub=7.98234003e+03
con1=wage_data>lb
con2=wage_data<ub
con=con1&con2
len(df_visa[con])

In [ ]:
con=ub-lb
con

In [ ]:
**Outlier imputation**

  - Method 1: 

     - lesstan lb outliers inputs with lb value

     - greaterthan ub outliers impute with ub value 

  - Method2: Every outlier replace with median

$cap: Winsorizaion 

In [ ]:
l=[]
for i in wage_data:
    if i<lb:
        l.append(lb)
    elif i>ub:
        l.append(ub)
    else:
        l.append(i)
df_visa['wage_data_new']

In [ ]:
import matplotlib.pyplot as plt

d = df_visa['wage_data_new']

plt.figure(figsize=(10, 5))
plt.subplot(2, 1, 1)
plt.hist(d, bins=30)
plt.title("Histogram")

plt.subplot(2, 1, 2)
plt.boxplot(d, vert=False)
plt.title("Boxplot")

plt.tight_layout()
plt.show()


In [ ]:
d=outliersdf['wage_data_new']
plt.subplot(2,1,1).hist(d)
plt.subplot(2,1,2).boxplot(d,vert=False)
plt.show()

In [ ]:
d=outliersdf['prevailing_wage']
plt.subplot(2,2,1).hist(d)
plt.subplot(2,2,2).hist(wage_data,bins=40)
plt.subplot(2,2,3).boxplot(d,vert=False)
plt.subplot(2,2,4).boxplot(wage_data,vert=False)
plt.show()

In [ ]:
plt.subplot(2,2,2).boxplot(wage_data,vert=False)


**Imputs With  Meadian**